src/v/cluster/rm_stm.cc

```diff
@@ -1216,43 +1222,54 @@ rm_stm::replicate_tx(model::batch_identity bid, model::record_batch_reader br) {
     } else {
         // this is the first attempt in the tx, reset dedupe cache
         reset_seq(bid);
-    }
 
-    _mem_state.inflight[bid.pid] = _mem_state.inflight[bid.pid] + 1;
+        _mem_state.estimated[bid.pid] = _insync_offset;
+    }
 
-    if (!_mem_state.tx_start.contains(bid.pid)) {
-        _mem_state.estimated.emplace(bid.pid, _insync_offset);
+    auto expiration_it = _log_state.expiration.find(bid.pid);
+    if (expiration_it == _log_state.expiration.end()) {
+        vlog(_ctx_log.warn, "Can not find expiration info for pid:{}", bid.pid);
+        co_return errc::generic_tx_error;
     }
+    expiration_it->second.last_update = clock_type::now();
+    expiration_it->second.is_expiration_requested = false;
 
     auto r = co_await _c->replicate(
       synced_term,
       std::move(br),
-      raft::replicate_options(raft::consistency_level::leader_ack));
+      raft::replicate_options(raft::consistency_level::quorum_ack));
     if (!r) {
         if (_mem_state.estimated.contains(bid.pid)) {
             // an error during replication, preventin tx from progress
             _mem_state.expected.erase(bid.pid);
         }
+        if (_c->is_leader() && _c->term() == synced_term) {
+            co_await _c->step_down();
```

Comment:
nit: add a log? for debugging.

Reply:
why do we need to step down here ?

Reply:
it's easier to write correct code when we don't need to think about ambiguity. for example here after the failure the status of the operation is unknown; if we step down then the next leader replays all the records written by previous leader and thus avoid the ambiguity